src/Processors/Formats/Impl/JSONEachRowRowInputFormat.cpp

```diff
@@ -267,6 +237,37 @@ bool JSONEachRowRowInputFormat::readRow(MutableColumns & columns, RowReadExtensi
     return true;
 }
 
+bool JSONEachRowRowInputFormat::checkEndOfData(bool is_first_row)
+{
+    /// We consume , or \n before scanning a new row, instead scanning to next row at the end.
+    /// The reason is that if we want an exact number of rows read with LIMIT x
+    /// from a streaming table engine with text data format, like File or Kafka
+    /// then seeking to next ;, or \n would trigger reading of an extra row at the end.
+
+    /// Semicolon is added for convenience as it could be used at end of INSERT query.
+    if (!in->eof())
+    {
+        /// There may be optional ',' (but not before the first row)
+        if (!is_first_row && *in->position() == ',')
+            ++in->position();
+        else if (!data_in_square_brackets && *in->position() == ';')
+        {
+            /// ';' means the end of query (but it cannot be before ']')
+            allow_new_rows = false;
+            return true;
+        }
+        else if (data_in_square_brackets && *in->position() == ']')
+        {
+            /// ']' means the end of query
+            allow_new_rows = false;
+            return true;
+        }
+    }
+
+    skipWhitespaceIfAny(*in);
```

Comment:
This and the next line look haven't been in the previous code state.

Reply:
These lines were in previous code state in `readRow` method, but I forgot to delete them from there, so now we have code duplication. Fixed.